packages/cf-util-http/src/http.js

```diff
@@ -33,84 +25,136 @@ export function beforeSend(callback) {
+  const queryParams = [];
+  // Clones the input
+  if (Array.isArray(kvs) || kvs instanceof Map) {
+    for (let [k, v] of kvs) {
+      queryParams.push([k, v]);
+    }
+  } else {
+    for (let k in kvs) {
+      if (kvs.hasOwnProperty(k)) {
+        queryParams.push([k, kvs[k]]);
+      }
+    }
+  }
+  return queryParams
+    .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
+    .join('&');
+}
+
+function toHash(headers) {
+  const hash = {};
+  for (let [k, v] of headers) {
+    hash[k] = v;
+    hash[k.toLowerCase()] = v;
+  }
+  return hash;
+}
+
+function wrapResponse(headers, status, body, text, response) {
+  return {
+    headers,
+    status,
+    body: headers['content-type'] === 'application/json'
+      ? JSON.parse(text)
+      : text,
+    text,
+    response
+  };
+}
+
 /**
- * Perform an http request.
+ * Perform an http request. This method uses WHATWG's fetch API and returns a
+ * promise that **resolves on any HTTP status**, and rejects on anything else,
+ * such as a malformed URL or a network error.
+
+ * If a callback is provided, and `response.ok` is true, the node-style callback
+ * will be called with an undefined error and a hash that contains the headers
+ * key-value pairs, status code, deserializes body (JSON or plain-text only for
+ * now), the body text, and the original `Response` object. Otherwise the
+ * callback is called with a single argument containing the above hash as the
+ * error value.
+ *
+ * This function also returns a promise that contains the fetched response
+ * object.
  *
  * ```js
  * request('POST', '/posts', {
  *   body: {
  *     title: 'A New Post',
  *     content: 'Contents of the new post.'
  *   }
- * }, function(res) {
- *   console.log(res.body); // > { result: { id: 1, title: 'A New Post', content: 'Contents of the new post.' } }
- * }, function(err) {
- *   console.log(res.body); // > { errors: [{ message: 'Error!' }] }
+ * }).then(function(res) {
+ *   res.json().then(console.log); // > { result: { id: 1, title: 'A New Post', content: 'Contents of the new post.' } }
  * });
  * ```
  *
  * @param {String} method - GET/POST/PUT/PATCH/DELETE
  * @param {String} url
- * @param {Object} [opts]
- * @param {Object} [opts.parameters]
- * @param {Object} [opts.headers]
- * @param {Object} [opts.body]
- * @param {Function} [callback]
- * @returns {Function} Abort request.
+ * @param {Object} [opts] Any options fetch()'s second parameter accepts
+ * @param {Object|Array<Array>|Map} [opts.parameters] Query parameters to add to `url`
+ * @returns {Promise} Fetch promise
+ *
+ * @see [Global Fetch on MDN](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch)
+ * @see [WHATWG Fetch API Spec](https://fetch.spec.whatwg.org/)
  */
-export function request(method, url, opts, callback) {
-  opts = opts || {};
-
-  opts.method = method;
+export function request(method, url, opts = {}, callback) {
+  opts.method = method.toUpperCase();
```

Comment:
do we want a good default?
`opts.method = method ? method.toUpperCase() : 'GET';`